sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/http/policy/BearerTokenAuthenticationChallengePolicy.java

```diff
@@ -0,0 +1,164 @@
+                               .flatMap(b -> Mono.just(true));
+                }
+            }
+        }
+        return Mono.just(false);
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            return Mono.error(new RuntimeException("token credentials require a URL using the HTTPS protocol scheme"));
+        }
+        HttpPipelineNextPolicy nextPolicy = next.clone();
+
+        return onBeforeRequest(context)
+                   .then(next.process())
+                   .flatMap(httpResponse -> {
+                       String authHeader = httpResponse.getHeaderValue(WWW_AUTHENTICATE);
+                       if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+                           return onChallengeAsync(context, httpResponse).flatMap(retry -> {
+                               if (retry) {
+                                   return nextPolicy.process();
+                               } else {
+                                   return Mono.just(httpResponse);
+                               }
+                           });
+                       }
+                       return Mono.just(httpResponse);
+                   });
+    }
+
+    /**
+     * Get the {@link AccessTokenCache} holding the cached access tokens and the logic to retrieve and refresh
+     * access tokens.
+     *
+     * @return the {@link AccessTokenCache}
+     */
+    public AccessTokenCache getTokenCache() {
+        return cache;
+    }
+
+    private Mono<Void> authenticateRequest(HttpPipelineCallContext context, Supplier<Mono<AccessToken>> tokenSupplier,
+                                           boolean forceTokenRefresh) {
+        return cache.getToken(tokenSupplier, forceTokenRefresh)
+           .flatMap(token -> {
+               context.getHttpRequest().getHeaders().put(AUTHORIZATION_HEADER, BEARER + " " + token.getToken());
+               return Mono.empty();
+           });
+    }
+
+    List<AuthenticationChallenge> parseChallenges(String header) {
+        Pattern pattern = Pattern.compile(AUTHENTICATION_CHALLENGE_REGEX);
+        Matcher matcher = pattern.matcher(header);
+
+        List<AuthenticationChallenge> challenges = new ArrayList<>();
+        while (matcher.find()) {
+            challenges.add(new AuthenticationChallenge(matcher.group(1), matcher.group(2)));
+        }
+
+        return challenges;
+    }
+
+    Map<String, String> parseChallengeParams(String challengeParams) {
+        Pattern pattern = Pattern.compile(AUTHENTICATION_CHALLENGE_PARAMS_REGEX);
+        Matcher matcher = pattern.matcher(challengeParams);
+
+        Map<String, String> challengeParameters = new HashMap<>();
+        while (matcher.find()) {
+            challengeParameters.put(matcher.group(1), matcher.group(2));
+        }
+        return challengeParameters;
+    }
```

Comment:
Is this code similar to what is in https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/AuthorizationChallengeHandler.java? If so could we move it into a centralized location?

Reply:
looks different.